### ISSUE TITLE
HIP port

### ISSUE THREAD

**Issue by jglaser** [mphoward/neighbor#6]
In evaluating the scope for a HIP port, and as we discussed, the main difficulty will be doing without the floating point rounding modes `neighbor` currently takes advantage of in CUDA. HIP doesn't support those, at least on AMD platforms. And it doesn't look like there will be support anytime soon https://github.com/RadeonOpenCompute/ROCm-Device-Libs/issues/64

I noticed there are two different rounding transitions in the library, one between single precision float and double, which is a kind of mixed precision support, and one between float ant int, corresponding to the compression scheme. Is it necessary to do the intermediate conversion into float, or could one directly go between double and int, perhaps at no significant performance cost? On the other hand, CUDA also supports double precision math operations in non-default rounding modes, so lack of support for double could not have been the reason for going through float. 

Secondly, the main comparison operators for the bounding volumes currently take float arguments, using positive/negative infinity rounding where appropriate. Would it be feasible to introduce the 10-bit int data type (or the three 10-bit ints packed into a 32bit uint) as the main argument for those operators, which would then internally do their math in fixed precision, i.e. by evaluating ints? One could provide a convenience wrapper that does a `__float2int_r(u/d)`, which seems supported by HIP. Therefore one might be able to avoid float math operations with non-nearest rounding mode entirely.

@mphoward Could you comment on the feasibility of the changes outlined above? Do they seem too much work? I could give them a try.

**Comment by jglaser** [comment]
I did a short survey of header-only fixed point libraries.
https://github.com/has-castle/fixed_point
https://github.com/johnmcfarlane/cnl

It looks like neither of them exposes the rounding modes (towards +/- infinity) that we need. Though CNL could perhaps be patched.

**Comment by jglaser** [comment]
After investigating further, it should be possible to treat the mixed precision features separately from the 10bit integer compression. It should be possible to disable mixed precision through a preprocessor define, which is only used when compiling against HIP/AMD, and which would eliminate the need for all the `_rd`/`_ru` arithmetics. These would reduce to built in double precision arithmetics in that mode.

**Comment by jglaser** [comment]
I am a bit puzzled by the rounding of `float`s to `int`s in `lbvh_compress_ropes` though.

```
    // high bounds are encoded relative to the high of the box, always rounding down
    const float3 hi = tree.hi[idx];
    const uint3 hi_bin = make_uint3((unsigned int)floorf(__fmul_rd(__fsub_rd(tree_hi.x,hi.x),tree_bininv.x)),
                                    (unsigned int)floorf(__fmul_rd(__fsub_rd(tree_hi.y,hi.y),tree_bininv.y)),
                                    (unsigned int)floorf(__fmul_rd(__fsub_rd(tree_hi.z,hi.z),tree_bininv.z)));
    const unsigned int hi_bin3 = (hi_bin.x << 20) + (hi_bin.y << 10) + hi_bin.z;
```
shouldn't it round up for the upper bounding box bounds `int3 hi_bin`, so that it is conservative? Using `ceil`?

**Comment by jglaser** [comment]
oh, no, I see, it is relative to the box hi

**Comment by mphoward** [comment]
> I am a bit puzzled by the rounding of `float`s to `int`s in `lbvh_compress_ropes` though.
> 
> ```
>     // high bounds are encoded relative to the high of the box, always rounding down
>     const float3 hi = tree.hi[idx];
>     const uint3 hi_bin = make_uint3((unsigned int)floorf(__fmul_rd(__fsub_rd(tree_hi.x,hi.x),tree_bininv.x)),
>                                     (unsigned int)floorf(__fmul_rd(__fsub_rd(tree_hi.y,hi.y),tree_bininv.y)),
>                                     (unsigned int)floorf(__fmul_rd(__fsub_rd(tree_hi.z,hi.z),tree_bininv.z)));
>     const unsigned int hi_bin3 = (hi_bin.x << 20) + (hi_bin.y << 10) + hi_bin.z;
> ```
> 
> shouldn't it round up for the upper bounding box bounds `int3 hi_bin`, so that it is conservative? Using `ceil`?

This needs to be floor (and relative to the upper edge) in order to avoid overflow, which would happen with ceil.

Regarding the mixed precision, the design of the library is to separate the accelerator (`LBVH`) from the traversal (`LBVHTraverser`) so that different traversal schemes can be implemented. The mixed-precision in the `LBVH` is one of convenience for users supplying doubles, and so this could be simplified or removed. (Putting the onus on the caller to decide how to convert their fp64 objects into fp32.)

I had thought before about how the traversal could maybe used fixed point math throughout, which might actually be faster and use fewer instructions than at present. But, it would break some of the simplicity of the current design because the logic to compute fixed-point overlaps is for floats not ints. We could consider adapting this.

I had some ideas for fixes that might introduce the compatibility you're looking for with HIP. Let me look at your PR i ndetail, and I'll get back to you on it.

**Comment by mphoward** [comment]
OK, I looked through the code, and I think that use of `*_ru` and `*_rd` is *correct* but may not be *necessary*. Accepting round-to-even is probably enough for most use cases. If the library is redesigned to work on floats, promising only float accuracy, then we could get rid of alot of the compatibility layer you had to introduce. (We could probably add a tolerance to tests if necessary.)

Given the number of changes you are trying to introduce, could we schedule a call sometime this coming week or next to discuss and make a plan?

**Comment by jglaser** [comment]
Yeah, sure. Happy to talk via skype or bluejeans. Although I do think the compatibility layer is a good thing, since it consists only of defines which intercept CUDA-specific intrinsics.

**Comment by jglaser** [comment]
> Yeah, sure. Happy to talk via skype or bluejeans. Although I do think the compatibility layer is a good thing, since it consists only of defines which intercept CUDA-specific intrinsics.

That said, if we can get rid of the intrinsics entirely, even better.

**Comment by mphoward** [comment]
I'm looking through your PR now. It seems like--other than the major math rounding issue we discussed--the changes required to support HIP are pretty minimal. Do you think we could do these with a few inline / template functions in a `Runtime.h` header so that HIP does not become a hard dependency of this library? I'm thinking of putting common functions into the `neighbor::gpu` namespace like:

```cuda
namespace neighbor { namespace gpu {

#ifdef HIP
#include <hip/hip_runtime.h>
typedef hipStream_t stream_t;
typedef hipError_t error_t;
#else
#include <cuda_runtime.h>
typedef cudaStream_t stream_t;
typedef cudaError_t error_t;
#endif

// streams
inline error_t streamCreate(stream_t*);
inline error_t streamSynchronize(stream_t);
inline error_t streamDestroy(stream_t);

// memory (malloc and free will be handled in Memory.h)
inline error_t memsetAsync(void*, int, size_t, stream_t);

// kernel launch
inline error_t getDeviceProperties(/*copy call signature*/);

template<class Kernel>
unsigned int getMaxBlockSize();

template<class Kernel, class ...Args>
void launchKernel(stream_t, unsigned int /*numblocks*/, unsigned int /*blocksize*/, unsigned int /*sharedbytes*/, Args... args);
}}
```

References to cub vs. hipcub could be worked around in functions as required (headers being in ifdefs):
```c++
namespace cub = hipcub;
```
The different `thrust` call signatures would probably just use the ifdef in place, as you have it.

Does HIP support managed memory (I think no?)? This would require some ifdef in `Memory.h` or modification of unit tests, as the device memory is currently not explicitly copied into host arrays.

What do you think?

**Comment by jglaser** [comment]
That sounds like a very workable solution, especially if the goal is to minimize dependencies for `neighbor`, which should be header-only and mostly stand alone. Your suggestions would even be a good idea for HOOMD, even though the fraction of the runtime API used might be larger.

In addition, we would have to introduce a few generalized defines, such as

```
#ifdef HIP
#define GPU_DEVICE_COMPILE __HIP_DEVICE_COMPILE___ 
#elif defined(CUDA)
#define GPU_DEVICE_COMPILE __CUDA_ARCH__
#endif

#ifdef HIP
#define GPUCC __HIPCC__
#elif defined(CUDA)
#define GPUCC __NVCC__
#endif
 ```

Thinking ahead. We could implement this for `neighbor` first, as the scope will perhaps be smaller, and then evaluate what would be required to abstract HIP/CUDA inside HOOMD. @joaander what do you think? This was deliberately not part of the initial HIP port of HOOMD (https://github.com/glotzerlab/hoomd-blue/pull/541), as abstractions are already provided through HIP. Now that we have a working build with HIP, there would be an opportunity here to generalize it further and make HOOMD more independent of either library backend, for future compatibility and easier maintenance. It would require changes on the C++ code similar to those applied using `hipify-perl`, as well as a slight refactoring of the build system. I anticipate most part of the work would be identifying all necessary API calls that need to be abstracted. For the user, the impact (and also benefit) would be minimal, as they would only have to download one or two less `git` submodules.

**Comment by mphoward** [comment]
Adding a few defines like that would be smart and readable. I don't think I currently make use of them here because neighbor is meant to be compiled only by nvcc, but that could be pretty useful to have.

I agree it's a good idea to try this for neighbor first, as you propose, since it is using such a small fraction of the runtime library. It also keeps the dependencies to a minimum, which I would like for neighbor given some of the pains I've seen you going through to get a workable HIP build system. :-)

In some sense, this is duplicating one of the core functionalities of HIP, but it is necessary if you want to be able to compile for pure CUDA instead of HIP-CUDA!

If it turns out to be useful, then porting to hoomd would also be straightforward, at the cost of expanding the features that need to be wrapped. To the hoomd user, there is essentially no difference since you've already put in the work. To the developer setting up a new project using neighbor, this could be a pretty useful feature. It may also be helpful for plugin developers--if they port their plugins once to the runtime interface header (rather than the HIP interface), then it no longer matters what the hoomd developers do behind the scenes.

**Comment by joaander** [comment]
This looks like a great idea to me. It would be great not to have force a hard dependency on HIP in HOOMD.

**Comment by mphoward** [comment]
I can start this by mocking up a minimal `Runtime.h` to reimplement the existing CUDA calls, similarly to what I've done in PR #10. Then, maybe @jglaser can take a pass at compilation and testing?

**Comment by mphoward** [comment]
@jglaser there's a preliminary version open in PR #11 if you want to take a look

**Comment by jglaser** [comment]
It would be great if we had also had some automated, python-based code transformation tool (`dehipify`? `gpuify`?), which we could keep up to date as we add methods to `Runtime.h`. It should be able to understand CUDA and HIP and be able to transform them to `gpu` syntax (such as, `cudaMemcpy` -> `gpu::memcpy`, 'hipFuncGetAttributes` -> `gpu::funcGetAttributes`). It would neither be very sophisticated, nor very complete, but just facilitate the repeating tasks, aiming to cover 70-80% of the tedious conversions. It could also then supplied with HOOMD to aid with the conversion of plugins.

**Comment by jglaser** [comment]
Haven't looked much into it, but this deprecated tool could be starting point 
https://github.com/rocmarchive/pyHIPIFY/blob/master/hipify-python.py

**Comment by mphoward** [comment]
This is a good idea, although I do not have time to work on it now. The transformations are pretty simple  under the names I've taken, since you mostly replace either `cudaX` or `hipX` by `gpu::x`. The trickiest part is updating the kernel launch. I suppose defines could also be a bit annoying, but maybe we can do it in a way where this is also find-replace.

**Comment by mphoward** [comment]
Resolved by #12
